Fix treatment of dotted keys when unknown=INCLUDE

sirosen opened the pull request:
resolves #1506

When `unknown=INCLUDE`, dotted keys are currently unpacked into nested dicts.
i.e. `{"foo.bar": "baz"} => {"foo": {"bar": "baz"}}` when `foo.bar` is not a known fieldname.

With this fix in place, dotted keys are only unpacked during field loading, not when loading additional fields for `INCLUDE`


In #1506, @deckar01 suggested using Field to do this, but I'm not sure I understood properly. This seems simpler to me and sufficient? Is there maybe a test-case with a Dict or other more complex fields that would require more sophisticated treatment.

This is a bugfix, so there's an argument for just releasing it as-is. However, I considered adding an option to schema opts for controlling this behavior. Something like `allow_dotted_unknown = True` by default, and then pass that to `allow_dotted` in `set_value`. I don't think that's worth it? But I could easily be convinced if one of the maintainers feels strongly about being _really strictly_ backwards compatible.

sirosen committed:
Fix treatment of dotted keys when unknown=INCLUDE

When `unknown=INCLUDE`, dotted keys are currently unpacked into nested
dicts.
i.e. `{"foo.bar": "baz"} => {"foo": {"bar": "baz"}}` when `foo.bar` is
not a known fieldname.

With this fix in place, dotted keys are only unpacked during field
loading, not when loading additional fields for `INCLUDE`

resolves #1506

deckar01 commented on src/marshmallow/schema.py:
```diff
@@ -682,7 +682,12 @@ def _deserialize(
+                            typing.cast(typing.Dict, ret),
+                            key,
+                            value,
+                            allow_dotted=False,
```
I would bypass `set_value()` and just make the dict assignment like @adyouri [suggested](https://github.com/marshmallow-code/marshmallow/issues/1506#issuecomment-583871756) to minimize the complexity.

sirosen replied:
Ah, that sounds so much better! I feel silly now. 😅 

I'm doing that now, and I'm going to rename to `ret_l` and `ret_d` for the `many=True/False` branches, so that mypy is pacified and we don't need `typing.cast(typing.Dict, ret)[key] = value` (which works, but yuck...).

deckar01 commented:
> In #1506, @deckar01 suggested using Field to do this, but I'm not sure I understood properly.

The underlying issue is that only certain field/meta arguments are allowed to parse dotted syntax, but this bug is parsing the dotted syntax of unknown data keys, which is undocumented (and undesired) behavior. The other `set_value()` call I was referring to is actually another use case where data keys can be parsed as dotted syntax. This bug is hard to reach, because dots are not valid Python class attribute names, but I suspect there is a way to hit it, possibly when generating a schema from a dict. The alternate solution only really applied to that use case, because the responsibility could be pushed down to the field. That wouldn't work for unknown keys, which by definition don't have a field declared, so I think fixing `_deserialize()` is good.

deckar01 commented:
> But I could easily be convinced if one of the maintainers feels strongly about being really strictly backwards compatible.

I think this is safe for a patch release without a feature flag. I don't think SEMVER is intended to protect building on undefined behavior. :shipit:

sirosen committed:
Replace set_value(allow_dotted=False) with setitem

Rather than adding the feature for utils.set_value() to do this, just
do dict key-value assignment. However, this results in a difficult to
read expression: `typing.cast(typing.Dict, ret)[key]`, when all that
is really *meant* is `ret[key]`.

Because mypy assigns types to variables and does not allow those types
to be reassigned or altered -- excepting type narrowing on unions --
the easy way out is to use different names for the `many=True/False`
branches of `_deserialize`. The branches now use `ret_l` (l for list)
and `ret_d` (d for dict) and there is no shared variable named `ret`.
This removes casts of `ret` to a dict, and results in
`ret_d[key] = value` in this case, which is the simplest/most readable
form.

sirosen commented:
I've simplified it to just `ret_d[key] = value` and gotten rid of the changes in `set_value`. (I amended because I fussed with a type comment and broke py3.5 compat, so I reverted that.)

> I don't think SEMVER is intended to protect building on undefined behavior.

👍 That's how I see it, but if someone wanted to make the case I wasn't going to resist.

> I suspect there is a way to hit it [dotted attribute name for a field], possibly when generating a schema from a dict.

Yeah, I bet that would do it. I can `class A: pass ; setattr(A, "foo.bar", 1)` and `dir(A)` shows `foo.bar` just fine. You're not able to access such an attribute with the normal dot-based syntax, but `getattr` and other methods should work fine.
I'll see if I can add a test or two for this to confirm the behavior is right, or -- if there is a bug -- I might try to put together a fix.

sirosen commented:
I think we should merge this as-is. I looked into dotted attributes and it seems like something so niche and strange that I think we could hang our hats on "the behavior is undefined".

<details><summary>Here's a diff from me playing with this.</summary>

```diff
diff --git a/src/marshmallow/schema.py b/src/marshmallow/schema.py
index ea030c1..3800dd6 100644
--- a/src/marshmallow/schema.py
+++ b/src/marshmallow/schema.py
@@ -673,7 +673,14 @@ class Schema(base.SchemaABC, metaclass=SchemaMeta):
                 )
                 if value is not missing:
                     key = field_obj.attribute or attr_name
-                    set_value(ret_d, key, value)
+                    if (
+                        field_obj.data_key is None
+                        and field_obj.attribute is None
+                        and "." in key
+                    ):
+                        ret_d[key] = value
+                    else:
+                        set_value(ret_d, key, value)
             if unknown != EXCLUDE:
                 fields = {
                     field_obj.data_key if field_obj.data_key is not None else field_name
diff --git a/tests/test_deserialization.py b/tests/test_deserialization.py
index dda9c59..4f71e25 100644
--- a/tests/test_deserialization.py
+++ b/tests/test_deserialization.py
@@ -1745,6 +1745,11 @@ class TestSchemaDeserialization:
         data = MySchema().load({"foo": "hi", "bar.baz": "okay", "alpha.beta": "woah!"})
         assert data == {"foo": "hi", "bar": "okay", "alpha.beta": "woah!"}

+    def test_field_with_dotted_attribute_name(self):
+        MySchema = Schema.from_dict({"foo.bar": fields.Str()})
+        data = MySchema().load({"foo.bar": "hi"})
+        assert data == {"foo.bar": "hi"}
+

 validators_gen = (func for func in [lambda x: x <= 24, lambda x: 18 <= x])
```

</details>

It works... but I'm not sure if that toy example is a realistic use-case? At the very least, I think we can follow-up on it separately.

sloria reviewed (approved):
Sorry it's taken so long to get this merged.

Fix looks great! Thanks also for the typing cleanup. 💯

sloria committed:
Merge branch 'dev' into fix-auto-dot-transform

sloria committed:
Update changelog
